### ISSUE TITLE
Pass the variant itself to the productVariantPriceCalculationsStrategy

### ISSUE THREAD

**Issue by JamesLAllen** [vendure-ecommerce/vendure#2126]
**Is your feature request related to a problem? Please describe.**
I need the product variant to match its sku with my customer pricing library, where each customer gets unique prices for each sku based on their existing contracts.  Their unique pricing needs to be displayed before they add the product to the cart.  I can create my custom `ProductVariantPriceCalculationStrategy` however since only the price before tax is passed to `calculate` I'm unable to get the sku of the variant.

**Describe the solution you'd like**
Change `product-price-applicator` to pass the `variant`:
```
const { price, priceIncludesTax } = await productVariantPriceCalculationStrategy.calculate({
            inputPrice: channelPrice.price,
            taxCategory: variant.taxCategory,
            activeTaxZone,
            ctx,
            productVariant: variant
        });
        ```
 and the updated args:
 ```
 export interface ProductVariantPriceCalculationArgs {
    inputPrice: number;
    taxCategory: TaxCategory;
    activeTaxZone: Zone;
    ctx: RequestContext;
    productVariant?: ProductVariant;
}
```

**Describe alternatives you've considered**
I'm using `patch-package` for my current needs however this is a hack but I expect will also prove the concept and allow me to work out any issues before I make the PR.  There also may be other ways to achieve what I'm after without this approach as I'm still learning all the ins and outs of Vendure.

**Comment by michaelbromley** [comment]
Hi James,

Thanks for the suggestion. The main issue I can see with this is of performance. Although the ProductVariantPriceCalculationStrategy (PVPCS) and OrderItemPriceCalculationStrategy  (OIPCS) seem to do more or less the same thing, there is a major difference: OIPCS is called quite rarely, only when changes are made to an order, whereas PVPCS is called _very_ frequently - every time you get a variant price from a response. E.g. if you have list view of product variants, it will be called for every single variant on the list. It is also used in other places internally when performing other price calculations. If you study the default strategy closely you'll find that it almost never does any real async work or make any DB calls - and this is with good reason.

I feel that encouraging people to do async work with DB lookups in this strategy will be a recipe for poor performance. So the limited nature of the API is a good constraint in that regard.

With that being said, how would you then handle the requirement of showing the correct price-list price to the customer without requiring them to add it to the order?

I think a better way would be to use a plugin to define a new field on the `ProductVariant` GraphQL type, and then you can define a custom resolver to perform your lookup from the price list.

Once we have some more experience implementing price lists in this or some similar way, I'd like to make available an official price lists plugin to make the process easier. So your feedback on this would be valuable.

(relates to #1920 )

**Comment by JamesLAllen** [comment]
Right, that makes complete sense however on every price check the `calculate` method already has to await the `taxService` and `channelService` in moreso in the `applyChannelPriceAndTax` method which calls `calculate`.  

After taking a look at the way I currently solved this, I think it works but requires some customization and would be preferred if I didn't even have to create a custom `ProductVariantPriceCalculationStrategy` and risk breaking things when you change something with taxes or something else, but rather be able to put into the config a new `ProductVariantPriceCalculationBeforeTaxStrategy` that the default would be just a dummy, but pass the variant and ctx where we can do whatever we want with it, and then any pre-tax prices are run through this strategy first.

I don't believe most use-cases would require any additional database lookups, just access to the variant itself and custom fields that it contains.  I also think that any bulk pricing lookups would require a completely different endpoint to be optimized and a different strategy employed altogether to take any lookups out of the pricing and channel calculations when you're querying thousands of product's prices.  For storefront and pre-cart prices however this is not required and seems to be working quite well for my needs.  

Here's what I did, but I think I would prefer a new `ProductVariantPriceCalculationBeforeTaxStrategy` that allowed me to make calculations before it checked the taxService:

I used `patch-package` to pass the variant to the `calculate` method and added the variant to the `ProductVariantPriceCalculationArgs` definition.  No real big changes or additional DB queries in the base Vendure.

```
const { price, priceIncludesTax } = await productVariantPriceCalculationStrategy.calculate({
            variant,
            inputPrice: channelPrice.price,
            taxCategory: variant.taxCategory,
            activeTaxZone,
            ctx,
        });```

Then, I created a new `ProductVariantPriceCalculationStrategy` for my plugin but in order to make it easy for plugin consumers to add their own custom calculations I created a dummy method:

```
async calculatePriceBeforeTaxes(
    args: ProductVariantPriceCalculationArgs & { itemCost?: number; customerCost?: number; customPrice?: number },
  ): Promise<number> {
    return args.inputPrice
  }```

The reason I put in a dummy method is because in your default `ProductVariantPriceCalculationStrategy` you're checking the `taxService` and doing a bunch of calculations and I would prefer that anyone consuming a plugin not have to worry about keeping their application up to date whenever Vendure makes changes to tax and other systems, just add in the pre-tax calculations.

In my B2B plugin I do query the pricingLibrary for the activeCompany, so the `calculate` method is now making 2 db queries instead of the one.  So far this is acceptable for showing the prices to customers max 50 per page but I agree with you that it's a compounding impact for each query we make within this method.

In my parent project implementation I then created my new `ProductVariantPriceCalculationStrategy` that extends the plugin's `ProductVariantPriceCalculationStrategy` and overrode the `calculatePriceBeforeTaxes` with the actual calculation for the specific business requirements:

1. Some companies are grandfathered in for specific prices for specific skus (use the `customPrice` in the customer's pricing library if it exists for that specific sku).
2. Any variant sku that doesn't exist in the pricing library will get the price based on itemCost + a flat 20% markup.
3. Any variant sku that IS added to the customer's pricing library will get a discounted price based on the difference from the price they're paying to a competitor based on the delta of the itemCost.

I plan to support sibling variants based on the original discount delta, but for the MVP we are just going with specific SKU.

And here's the strategy that goes into my parent project's `vendure-config`:

```export class InvoCurePriceCalculationStrategy extends CompanyProductVariantPriceCalculationStrategy {
  async calculateBeforeTaxesPrice(
    args: ProductVariantPriceCalculationArgs & { itemCost?: number; customerCost?: number; customPrice?: number },
  ): Promise<number> {
    const { inputPrice, itemCost, customPrice } = args
    if (customPrice) return customPrice
    const defaultMarkupPercentage = 0.2
    let outputPrice = inputPrice
    // set the default price to cost + markup

    if (itemCost) {
      outputPrice = itemCost + itemCost * defaultMarkupPercentage
    }
    const customerCost = args.customerCost || outputPrice
    // get the delta between cost and customerCost
    const deltaPercentage = itemCost && customerCost ? (itemCost - customerCost) / customerCost : 0
    // set a default markdown percentage
    let markdownPercentage = 0
    // if the delta is negative, set the markdown percentage based on the difference
    if (deltaPercentage < -0.1) markdownPercentage = 0.05
    if (deltaPercentage < -0.2) markdownPercentage = 0.1
    if (deltaPercentage < -0.4) markdownPercentage = 0.2
    if (deltaPercentage < -0.6) markdownPercentage = 0.3
    outputPrice = customerCost - customerCost * markdownPercentage
    return outputPrice
  }
}
```

**Comment by michaelbromley** [comment]
> Right, that makes complete sense however on every price check the calculate method already has to await the taxService and additionally in the applyChannelPriceAndTax method which calls calculate it also queries the channelService.

These are both cached so typically they don't actually hit the database.

Thanks for providing all that detail on your use-case. I'm not totally clear yet on how the `ProductVariantPriceCalculationBeforeTaxStrategy` would work or differs from the existing strategy - since variant prices can be stored with or without tax included (depending on channel settings), it is not straightforward to know whether a price is "before" tax, without taking into account current channel settings. 

Another thing that's worth noting is that there are changes coming to the way we handle prices in v2. There is a new [ProductVariantPriceSelectionStrategy](https://github.com/vendure-ecommerce/vendure/blob/efe23d127ffcc7d51c9ea4c42821842f70bb4957/packages/core/src/config/catalog/product-variant-price-selection-strategy.ts#L6) because now a variant can have multiple prices on a single channel - the intention being to be able to support multiple currencies on a single channel, but perhaps the concept can also be used to support use cases like yours somehow. I'm afraid this part is not yet documented, but I'll be working on that in the coming weeks.

**Comment by JamesLAllen** [comment]
I think the only reason I might suggest `ProductVariantPriceCalculationBeforeTaxStrategy` would be to decouple the price calculation from the tax portion, so that people wouldn't have to mess with the tax calculation unless they wanted to.  Currently I kept your original tax calculations in my strategy.

Since they are both cached, would my implementation also be cached as well?

**Comment by dlhck** [comment]
@michaelbromley we should revisit that with #3043

**Comment by dlhck** [comment]
This feature request has been moved to GitHub Discussions for community voting.
Please upvote the discussion if you would like to see this implemented: https://github.com/vendurehq/vendure/discussions/4289
